promql/prettier/prettier.go

```diff
@@ -0,0 +1,480 @@
+		}
+		result += item.Val
+		if p.pd.containsGrouping {
+			p.pd.containsGrouping = false
+			result += p.pd.newLine()
+		}
+	case parser.LEFT_BRACE:
+		result += item.Val
+		if nodeSplittable {
+			result += p.pd.newLine()
+			p.pd.insideMultilineBraces = true
+		}
+	case parser.RIGHT_BRACE:
+		if p.pd.insideMultilineBraces {
+			if items[index-1].Typ != parser.COMMA {
+				// Edge-case: if the labels are multi-line split, but do not have
+				// a pre-applied comma.
+				result += "," + p.pd.newLine() + p.pd.pad(nodeInfo.baseIndent(items[index]))
+			}
+			p.pd.insideMultilineBraces = false
+		}
+		result += item.Val
+	case parser.IDENTIFIER:
+		if p.pd.insideMultilineBraces {
+			result += p.pd.pad(1) + item.Val
+		} else {
+			result += item.Val
+		}
+	case parser.STRING, parser.NUMBER:
+		result += item.Val
+	case parser.EQL, parser.EQL_REGEX, parser.NEQ, parser.NEQ_REGEX, parser.DURATION, parser.COLON,
+		parser.LEFT_BRACKET, parser.RIGHT_BRACKET:
+		result += item.Val
+	case parser.ADD, parser.SUB, parser.DIV, parser.GTE, parser.GTR, parser.LOR, parser.LAND,
+		parser.LSS, parser.LTE, parser.LUNLESS, parser.MOD, parser.POW:
+		if p.pd.containsGrouping {
+			result += p.pd.newLine() + p.pd.pad(baseIndent) + item.Val + " "
+		} else if nodeSplittable && !p.pd.immediateScalar {
+			result += p.pd.newLine() + p.pd.pad(baseIndent) + item.Val + p.pd.newLine()
+		} else {
+			result += " " + item.Val + " "
+		}
+		if p.pd.immediateScalar {
+			p.pd.immediateScalar = false
+		}
+	case parser.WITHOUT, parser.BY, parser.IGNORING, parser.BOOL, parser.GROUP_LEFT, parser.GROUP_RIGHT,
+		parser.OFFSET, parser.ON:
+		result += item.Val
+		if item.Typ == parser.BOOL {
+			result += p.pd.newLine()
+		}
+	case parser.COMMA:
+		result += item.Val
+		if nodeSplittable {
+			result += p.pd.newLine()
+		} else {
+			result += " "
+		}
+	case parser.COMMENT:
+		if p.pd.isNewLineApplied {
+			result += p.pd.previousBaseIndent + item.Val + p.pd.newLine()
+		} else {
+			result += item.Val + p.pd.newLine()
+		}
+	}
+	if index+1 == len(items) {
+		return result, nil
+	}
+	ptmp, err := p.prettify(items, index+1, result)
+	if err != nil {
+		return result, errors.Wrap(err, "prettify")
+	}
+	return ptmp, nil
+}
+
+// refreshLexItems refreshes the contents of the lex slice and the properties
+// within it. This is expected to be called after sorting the lexItems in the
+// pre-format checks.
+func (p *Prettier) refreshLexItems(items []parser.Item) []parser.Item {
+	return p.lexItems(p.expressionFromItems(items))
+}
+
+// expressionFromItems returns the raw expression from slice of items.
+// This is mostly used in re-ordering activities where the position of
+// the items change during sorting in the sortItems.
+// This function also standardizes the expression without any spaces,
+// so that the length determined by .String() is done purely on the
+// character length with standard indent.
+func (p *Prettier) expressionFromItems(items []parser.Item) string {
```

Comment:
The return type here is not an expression.

Reply:
The function aims to return stringified expression, not `parser.Expr`. I guess, the problem lies in the naming.

Reply:
But a single item often isn't an expression. So yes, the name is confusing.

Reply:
Renamed to `stringifiedExpressionFromItems(items []parser.Items)`. Hence, resolving.